### PR TITLE
Mark vars as do not enreg earlier in minopts.

### PR DESCRIPTION
**Pull request by sandreenko**
and improve morph block init/copy a little bit.

There is a known problem that we start asking `is it already known to be not on register?` before the value is finalized during morph (block copy, should we do field by field) and lowering (do we need bitcast? can we mark the source as contained?) .
I have an markdown description of it in 6.0 planning.

For now, just mark them as `doNotEnreg` earlier when enregistration is disabled. Give us small improvements on minopts methods and prevent some regressions with future changes.

```
coreclr_tests.pmi.windows.x64.checked.mch
Total bytes of delta: -4
1 total methods with Code Size differences (1 improved, 0 regressed)

coreclr_tests.pmi.Linux.arm64.checked.mch
Total bytes of delta: -8
2 total methods with Code Size differences (2 improved, 0 regressed)

libraries_tests.pmi.Linux.arm64.checked.mch
Total bytes of delta: -28
5 total methods with Code Size differences (5 improved, 0 regressed)

benchmarks.run.Linux.x64.checked.mch
Total bytes of delta: -3
1 total methods with Code Size differences (1 improved, 0 regressed)

coreclr_tests.pmi.Linux.x64.checked.mch
Total bytes of delta: -4 
1 total methods with Code Size differences (1 improved, 0 regressed)

benchmarks.run.windows.arm64.checked.mch
Total bytes of delta: -8
1 total methods with Code Size differences (1 improved, 0 regressed)

coreclr_tests.pmi.windows.arm64.checked.mch
Total bytes of delta: -8
2 total methods with Code Size differences (2 improved, 0 regressed)

libraries_tests.pmi.windows.arm64.checked.mch
Total bytes of delta: -28
5 total methods with Code Size differences (5 improved, 0 regressed)

coreclr_tests.pmi.windows.x86.checked.mch
Total bytes of delta: -4
1 total methods with Code Size differences (1 improved, 0 regressed)
```

+no diffs improvements of morph block init/copy.

### CODE REVIEW

File: src/coreclr/jit/morphblock.cpp [hunk 1]
```diff
@@ -1031,6 +1006,40 @@ void MorphCopyBlockHelper::MorphStructCases()
         m_result                 = CopyFieldByField();
         m_transformationDecision = BlockTransformation::FieldByField;
     }
+
+    // Mark the dest/src structs as DoNotEnreg when they are not being fully referenced as the same type.
+    //
+    if (!m_dstDoFldAsg && (m_dstVarDsc != nullptr) && !m_dstSingleLclVarAsg)
```
Comment (sandreenko) [inline src/coreclr/jit/morphblock.cpp:1012]: after this change init/copy like:

               [000015] -ACXG-------              *  ASG       struct (copy)
               [000013] D----+-N----              +--*  LCL_VAR   struct<S, 8> V01 loc0
               [000012] --CXG+------              \--*  CALL      struct TestStructFields.Program.GetS1
don't require doNotEnreg flag. The condition was moved after we do m_comp->fgMorphBlockOperand to work after OBJ(ADDR(LCL)) collapse.

File: src/coreclr/jit/lclvars.cpp [hunk 2]
```diff
@@ -3467,7 +3473,7 @@ void Compiler::lvaSortByRefCount()
             assert(varDsc->lvType != TYP_STRUCT ||
                    varDsc->lvDoNotEnregister); // For structs, should have set this when we set lvAddrExposed.
         }
-        else if (varTypeIsStruct(varDsc))
+        if (varTypeIsStruct(varDsc))
```
Comment (sandreenko) [inline src/coreclr/jit/lclvars.cpp:3476]: the idea of `if/else` was that each block was setting `lvaSetVarDoNotEnregister` and there was no need to check the rest, now we have blocks that don't set it, so we need to check each.

**Comment by sandreenko** [comment]
PTAL @BruceForstall @dotnet/jit-contrib , another preparation for structs in registers, small changes and diffs.

**Review by BruceForstall — APPROVED** [review]

**Comment by kunalspathak** [comment]
@sandreenko - We noticed several regressions possibly related to this change. Do you mind going through some of them to confirm if they are indeed from this change?
- https://github.com/DrewScoggins/performance-2/issues/7139
- https://github.com/DrewScoggins/performance-2/issues/7142
- https://github.com/DrewScoggins/performance-2/issues/7143

**Comment by EgorBo** [comment]
Simplest repro:

```
cd src\benchmarks\micro
dotnet run -c Release -f net5.0 -- --coreRun C:\prj\runtime\artifacts\tests\coreclr\windows.x64.Release\Tests\Core_Root\corerun.exe --filter "*IniArray.Test*" 
```
It prints around `179ms`, then if I revert changes from this PR by hands and re-run - it prints `45ms`. The benchmark itself seems to have the same codegen so I have no clue what's going on...

**Comment by sandreenko** [comment]
> @sandreenko - We noticed several regressions possibly related to this change. Do you mind going through some of them to confirm if they are indeed from this change?

Sure, looking at them now, thanks for reporting it.

**Comment by sandreenko** [comment]
> ```
> dotnet run -c Release -f net5.0 -- --coreRun C:\prj\runtime\artifacts\tests\coreclr\windows.x64.Release\Tests\Core_Root\corerun.exe --filter "*IniArray.Test*" 
> ```

Hm, don't see this:
```
D:\Sergey\git\runtime\.dotnet\dotnet.exe run -c Release -f net5.0 --  --cli D:\Sergey\git\runtime\.dotnet\dotnet.exe  --filter "*IniArray.Test*" --coreRun "D:\Sergey\git\runtime\artifacts\tests\coreclr\windows.x64.Release\Tests\Core_Root_base\corerun.exe"

|   Method |     Mean |    Error |   StdDev |   Median |      Min |      Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------- |---------:|---------:|---------:|---------:|---------:|---------:|------:|------:|------:|----------:|
| IniArray | 42.86 ms | 0.826 ms | 0.690 ms | 42.56 ms | 42.33 ms | 44.75 ms |     - |     - |     - |      88 B |

D:\Sergey\git\runtime\.dotnet\dotnet.exe run -c Release -f net5.0 --  --cli D:\Sergey\git\runtime\.dotnet\dotnet.exe  --filter "*IniArray.Test*" --coreRun "D:\Sergey\git\runtime\artifacts\tests\coreclr\windows.x64.Release\Tests\Core_Root_diff\corerun.exe"
|   Method |     Mean |    Error |   StdDev |   Median |      Min |      Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------- |---------:|---------:|---------:|---------:|---------:|---------:|------:|------:|------:|----------:|
| IniArray | 42.57 ms | 0.175 ms | 0.136 ms | 42.58 ms | 42.35 ms | 42.80 ms |     - |     - |     - |      88 B |
```

I will check other regressions.

**Comment by EgorBo** [comment]
@sandreenko I've just checked on a clean repo - still reproduces for me

**Comment by EgorBo** [comment]
@sandreenko managed to get the diff between two coreruns: https://www.diffchecker.com/qh9EOTD4

**Comment by sandreenko** [comment]
> @sandreenko managed to get the diff between two coreruns: https://www.diffchecker.com/qh9EOTD4

Thank you, now I see them, I had a system-wide tiering disabled that I forgot about.

**Pull request merged**